### PR TITLE
Feature/asio

### PR DESCRIPTION
**Pull request by albertogpz**
This draft pull request includes the rebase of develop into the feature/asio branch and the solution to the hangs and aborts that were present in the branch in the integration-tests (see GEODE-8318).

All test cases are passing now, even the IPV6 one, without the need to configure IPv6 in the CMakeLists.txt file.

The code is ready for review.

### CODE REVIEW

**Comment by pdxcodemonkey** [comment]
This is failing one of our "acceptance" tests, `SNITest.dropSNIProxy`.  This is a negative test case, so I think it's expecting a Geode exception and now it's getting the following:
```
terminate called after throwing an instance of 'boost::wrapexcept<boost::system::system_error>'
  what():  remote_endpoint: Transport endpoint is not connected
```
Command line for CI running acceptance tests is the following:
```
cd /home/geode/build/cppcache/acceptance-test && ctest -C Debug --output-on-failure --output-log test.log --timeout 600
```
You can probably repro with just this:
```
cd /home/geode/build/cppcache/acceptance-test && ctest -C Debug -R SNITest.dropSNIProxy --timeout 600
```

**Comment by pdxcodemonkey** [comment]
Outstanding work, BTW!  As advertised, _everything_ else appears to be passing.  This is a huge win, thanks so much!

**Comment by albertogpz** [comment]
> This is failing one of our "acceptance" tests, `SNITest.dropSNIProxy`. This is a negative test case, so I think it's expecting a Geode exception and now it's getting the following:
> 
> ```
> terminate called after throwing an instance of 'boost::wrapexcept<boost::system::system_error>'
>   what():  remote_endpoint: Transport endpoint is not connected
> ```
> 
> Command line for CI running acceptance tests is the following:
> 
> ```
> cd /home/geode/build/cppcache/acceptance-test && ctest -C Debug --output-on-failure --output-log test.log --timeout 600
> ```
> 
> You can probably repro with just this:
> 
> ```
> cd /home/geode/build/cppcache/acceptance-test && ctest -C Debug -R SNITest.dropSNIProxy --timeout 600
> ```

I forgot about the acceptance-tests :-).
I have fixed the issue in the new commit.

**Comment by pdxcodemonkey** [comment]
Down to one failed test on a couple of platforms now - so close!

```
testXmlCacheCreationWithPools
```
Is timing out on Windows and RHEL7.

FYI, I'm out this week for the US holiday, so will be of limited, if any, help.  @mmartell & @mreddington can follow up as needed, I'll ping them via DM.

**Comment by albertogpz** [comment]
> Down to one failed test on a couple of platforms now - so close!
> 
> ```
> testXmlCacheCreationWithPools
> ```
> 
> Is timing out on Windows and RHEL7.
> 
> FYI, I'm out this week for the US holiday, so will be of limited, if any, help. @mmartell & @mreddington can follow up as needed, I'll ping them via DM.

I managed to fix the test on windows thanks to @gaussianrecurrence . I hope it is also fixed under RHEL7.

**Comment by albertogpz** [comment]
> > Down to one failed test on a couple of platforms now - so close!
> > ```
> > testXmlCacheCreationWithPools
> > ```
> > 
> > 
> > Is timing out on Windows and RHEL7.
> > FYI, I'm out this week for the US holiday, so will be of limited, if any, help. @mmartell & @mreddington can follow up as needed, I'll ping them via DM.
> 
> I managed to fix the test on windows thanks to @gaussianrecurrence . I hope it is also fixed under RHEL7.

Anyway, I have a question for @mreddington . The test case failed due to the addition of the following commit:
d2a307d80 cppcache/src/TcrEndpoint.cpp     (Matthew Reddington 2020-05-14 17:36:32 -0700 1181)   m_cacheImpl->doIfDestroyNotPending([&]()
 {

What is the reason for this change? I have not seen it in the development branch so I wonder if it is really necessary.

**Comment by pdxcodemonkey** [comment]
This is passing on Windows now, and I have at least an example of it working on RHEL7 Release build.  Our Debug and Coverage builds of RHEL7 are now timing out on a test, though, which has been a problem in the past.  Here's the tail end of the output, which looks kind of strange to me.

```
[       OK ] PartitionRegionOpsTest.putPartitionedRegionWithRedundancyServerGoesDownSingleHop (60480 ms)
[----------] 1 test from PartitionRegionOpsTest (60480 ms total)

[----------] Global test environment tear-down
[==========] 1 test from 1 test suite ran. (60481 ms total)
[  PASSED  ] 1 test.
[debug 2020/11/25 21:03:47.396293 UTC heavy-lifter-80e74939-d075-5450-9833-2e242c016911:1522 139806642039104] ThinClientPoolDM::destroy...
[debug 2020/11/25 21:03:47.396528 UTC heavy-lifter-80e74939-d075-5450-9833-2e242c016911:1522 139806642039104] Internal fair queue size while closing is 0
[debug 2020/11/25 21:03:47.396541 UTC heavy-lifter-80e74939-d075-5450-9833-2e242c016911:1522 139806642039104] ConnectionQueue::close( ): queue closed 
[debug 2020/11/25 21:03:47.396559 UTC heavy-lifter-80e74939-d075-5450-9833-2e242c016911:1522 139806642039104] recursive lock: CacheImpl::doIfDestroyNotPending


0% tests passed, 1 tests failed out of 1

Total Test time (real) = 650.16 sec

The following tests FAILED:
	 27 - PartitionRegionOpsTest.putPartitionedRegionWithRedundancyServerGoesDownSingleHop (Timeout)
Errors while running CTest
```

**Comment by albertogpz** [comment]
> This is passing on Windows now, and I have at least an example of it working on RHEL7 Release build. Our Debug and Coverage builds of RHEL7 appear to now be timing out on a test, though, which has been a problem in the past. Here's the tail end of the output, which looks kind of strange to me.
> 
> ```
> [       OK ] PartitionRegionOpsTest.putPartitionedRegionWithRedundancyServerGoesDownSingleHop (60480 ms)
> [----------] 1 test from PartitionRegionOpsTest (60480 ms total)
> 
> [----------] Global test environment tear-down
> [==========] 1 test from 1 test suite ran. (60481 ms total)
> [  PASSED  ] 1 test.
> [debug 2020/11/25 21:03:47.396293 UTC heavy-lifter-80e74939-d075-5450-9833-2e242c016911:1522 139806642039104] ThinClientPoolDM::destroy...
> [debug 2020/11/25 21:03:47.396528 UTC heavy-lifter-80e74939-d075-5450-9833-2e242c016911:1522 139806642039104] Internal fair queue size while closing is 0
> [debug 2020/11/25 21:03:47.396541 UTC heavy-lifter-80e74939-d075-5450-9833-2e242c016911:1522 139806642039104] ConnectionQueue::close( ): queue closed 
> [debug 2020/11/25 21:03:47.396559 UTC heavy-lifter-80e74939-d075-5450-9833-2e242c016911:1522 139806642039104] recursive lock: CacheImpl::doIfDestroyNotPending
> 
> 
> 0% tests passed, 1 tests failed out of 1
> 
> Total Test time (real) = 650.16 sec
> 
> The following tests FAILED:
> 	 27 - PartitionRegionOpsTest.putPartitionedRegionWithRedundancyServerGoesDownSingleHop (Timeout)
> Errors while running CTest
> ```

Interesting... @gaussianrecurrence told me that he had seen a deadlock in the feature/asio branch once related to that mutex but had not been able to reproduce it again. We looked together at the code today and it appeared that a deadlock is possible after the change included by @mreddington about which I him asked above. We could try removing the call to CacheImpl::doIfDestroyNotPending in TcrEndpoint::closeConnections() to see if the timeout does not occur.

**Comment by mreddington** [comment]
@albertogpz Despite it's terrible name, `doIfDestroyNotPending` ends a race condition where code tries to access the cache after it was destroyed. It's a shame we apparently need to set max connections which is only available from the distributed system when we close connections. This could cause a segfault during client shutdown.

**Comment by albertogpz** [comment]
> @albertogpz Despite it's terrible name, `doIfDestroyNotPending` ends a race condition where code tries to access the cache after it was destroyed. It's a shame we apparently need to set max connections which is only available from the distributed system when we close connections. This could cause a segfault during client shutdown.

Thanks @mreddington .
@pdxcodemonkey , I have pushed a new commit to see if removing the need to use that method (and the use itself) in `TcrEndpoint::closeConnections()`, fixes the possible deadlock. I do not have a RHEL7 environment so I cannot verify if this change fixes anything. Could you please test it?

**Comment by gaussianrecurrence** [comment]
> @albertogpz Despite it's terrible name, `doIfDestroyNotPending` ends a race condition where code tries to access the cache after it was destroyed. It's a shame we apparently need to set max connections which is only available from the distributed system when we close connections. This could cause a segfault during client shutdown.

Problem is that using doIfDestroyNotPending has undefined behaviour, as you are accessing the CacheImpl status equally. If it happens that the cache is removed and the mem page on which the cache was gets overwritten, doIfDestroyNotPending could try to execute the lambda.

The real issue here is that whenever ThinClientPoolDM::destroy is called, m_endpoints is not cleared
By the time ThinClientPoolDM::~ThinClientPoolDM is called, m_endpoints is to be destroyed, decrementing the reference counter to each of the container items.
If it happens that the container is holding an object which this was its latest reference, TcrEndpoint::TcrEndpoint will be called leading to undefined behaviour.

The real solution here is to make sure that before a pool destructor is called, all the references to the endpoints are cleared out.

**Comment by pdxcodemonkey** [comment]
Our CI didn't run builds after the last commit, because it has a `clang-format` issue:

```
@@ -140,7 +140,7 @@ class NamingContextImpl : virtual public NamingContext {
       LOGCOORDINATOR(func);
       LOGCOORDINATOR("Dump follows:");
       dump();
-      throw -1;
+      throw - 1;
     }
     return result;
   }
```
If you fix/push this, it should kick off a new set of builds.  Thanks!

**Comment by albertogpz** [comment]
> Our CI didn't run builds after the last commit, because it has a `clang-format` issue:
> 
> ```
> @@ -140,7 +140,7 @@ class NamingContextImpl : virtual public NamingContext {
>        LOGCOORDINATOR(func);
>        LOGCOORDINATOR("Dump follows:");
>        dump();
> -      throw -1;
> +      throw - 1;
>      }
>      return result;
>    }
> ```
> 
> If you fix/push this, it should kick off a new set of builds. Thanks!

Done. Strangely, it was cmake the one who changed that line in the first place to `throw -1`.

**Comment by pdxcodemonkey** [comment]
This is green in CI now - can you please rebase the branch onto the latest `develop`?  We'll review it quickly and get it merged.

**Comment by pdxcodemonkey** [comment]
FWIW, a rebase is liable to be unpleasant, from what I saw when attempting it.  I saved off the contents of a merge commit into `develop` as a stash, applied it to a new branch, and came up with this:  https://github.com/apache/geode-native/compare/develop...pdxcodemonkey:feature/asio-latest.  Hope this helps.

**Comment by albertogpz** [comment]
> This is green in CI now - can you please rebase the branch onto the latest `develop`? We'll review it quickly and get it merged.

Done in the latest force push. Not sure why the checks are failing with merge conflicts. Everything looked correct in my local branch.

**Comment by pdxcodemonkey** [comment]
Your PR is still set up to merge from your fork into the apache `feature/asio` branch, rather than `develop`.  I think if you target `develop` the conflicts will be gone.

**Comment by albertogpz** [comment]
> Your PR is still set up to merge from your fork into the apache `feature/asio` branch, rather than `develop`. I think if you target `develop` the conflicts will be gone.

You are right. I changed that and now I get errors in LGTM analysis but I do not understand why.

**Comment by albertogpz** [comment]
> > Your PR is still set up to merge from your fork into the apache `feature/asio` branch, rather than `develop`. I think if you target `develop` the conflicts will be gone.
> 
> You are right. I changed that and now I get errors in LGTM analysis but I do not understand why.

There was a merge commit from develop that I should not have added.
Now it is fixed and ready for review.

**Review by pdxcodemonkey — CHANGES_REQUESTED** [review]
Looks great, thanks again!  Just a couple of small edits, and a couple of questions to answer, then you're good to go.  I'll try to get this merged on Monday if you're ready.

File: cppcache/src/Connector.hpp [hunk 1]
```diff
@@ -37,24 +37,28 @@ constexpr std::chrono::milliseconds DEFAULT_READ_TIMEOUT = DEFAULT_TIMEOUT;
 
 constexpr std::chrono::milliseconds DEFAULT_WRITE_TIMEOUT = DEFAULT_TIMEOUT;
 
-class Connector {
+class APACHE_GEODE_EXPORT Connector {
```
[the remark this answers is not in the document]
Comment (albertogpz) [reply cppcache/src/Connector.hpp:40]: This was already there in the version of the branch I took as base. I have removed the export as I do not see the need.

File: cppcache/src/Connector.hpp [hunk 2]
```diff
@@ -63,45 +67,88 @@ class Connector {
...
+   * @return  The number of bytes written. Don't get confused: this is not the
+   * number of elements in the array written.
+   * @exception GeodeIOException
    */
-  virtual uint16_t getPort() = 0;
+  template <typename T, size_t size>
+  size_t receive(T (&array)[size], std::chrono::milliseconds timeout) {
+    return receive_nothrowiftimeout(reinterpret_cast<char *>(array),
+                                    sizeof(T) * size, timeout);
+  }
```
[the remark this answers is not in the document]
Comment (albertogpz) [reply cppcache/src/Connector.hpp:149]: Class not finally exported.

File: cppcache/src/DistributedSystem.hpp [hunk 3]
```diff
@@ -108,7 +107,6 @@ class APACHE_GEODE_EXPORT DistributedSystem {
 
   friend class CacheRegionHelper;
   friend class DistributedSystemImpl;
-  friend class TcrConnection;
 };
```
[the remark this answers is not in the document]
Comment (albertogpz) [reply cppcache/src/DistributedSystem.hpp:110]: Done by someone else in the version of the branch I took as base.

File: cppcache/src/TcrConnection.hpp [hunk 4]
```diff
@@ -120,43 +109,12 @@ class TcrConnection {
    * @param     ports     List of local ports for connections to endpoint
    */
   bool initTcrConnection(
-      TcrEndpoint* endpointObj, const char* endpoint,
+      std::shared_ptr<TcrEndpoint> endpointObj,
```
[the remark this answers is not in the document]
Comment (albertogpz) [reply cppcache/src/TcrConnection.hpp:112]: Done by someone else in the version of the branch I took as base.

File: defaultSystem/geode.properties [hunk 5]
```diff
@@ -28,8 +28,8 @@
 #
 ## Log file config
 #
-#log-file=geode_native.log
-#log-level=config
+log-file=geode_native.log
+log-level=debug
```
[the remark this answers is not in the document]
Comment (albertogpz) [reply defaultSystem/geode.properties:32]: Done by someone else in the version of the branch I took as base. I have left it as it was originally.

File: tests/cpp/fwklib/TcpIpc.cpp [hunk 6]
```diff
@@ -106,21 +106,6 @@ void TcpIpc::init(int32_t sockBufferSize) {
   ACE_OS::signal(SIGPIPE, SIG_IGN);  // Ignore broken pipe
```
[the remark this answers is not in the document]
Comment (albertogpz) [reply tests/cpp/fwklib/TcpIpc.cpp:106]: I guess you are referring to:
ssl_keys/server_keys/locator-maeve-keystore.jks
ssl_keys/server_keys/server-clementine-keystore.jks
ssl_keys/server_keys/server-dolores-keystore.jks
ssl_keys/server_keys/truststore.jks

They also came in the version of the branch I took as base. I have removed them.

File: cppcache/src/PoolFactory.cpp [hunk 7]
```diff
@@ -172,14 +172,14 @@ PoolFactory& PoolFactory::setServerGroup(std::string group) {
...
-  addCheck(host, port);
+  //  addCheck(host, port);
   m_attrs->addLocator(host, port);
   m_addedServerOrLocator = true;
   return *this;
 }
 
 PoolFactory& PoolFactory::addServer(const std::string& host, int port) {
-  addCheck(host, port);
+  //  addCheck(host, port);
```
[the remark this answers is not in the document]
Comment (albertogpz) [reply cppcache/src/PoolFactory.cpp]: Done by someone else in the version of the branch I took as base. Removed the commented out code.

**Comment by pdxcodemonkey** [comment]
Well I guess we know why Connector needs to be APACHE_GEODE_EXPORT now:

```
c:\nativeclient\cppcache\src\TcpConn.hpp(33): error C2220: warning treated as error - no 'object' file generated (compiling source file C:\nativeclient\cppcache\src\TcpConn.cpp) [C:\build\cppcache\shared\apache-geode.vcxproj]
c:\nativeclient\cppcache\src\TcpConn.hpp(33): warning C4275: non dll-interface class 'apache::geode::client::Connector' used as base for dll-interface class 'apache::geode::client::TcpConn' (compiling source file C:\nativeclient\cppcache\src\TcpConn.cpp) [C:\build\cppcache\shared\apache-geode.vcxproj]
  c:\nativeclient\cppcache\src\Connector.hpp(40): note: see declaration of 'apache::geode::client::Connector' (compiling source file C:\nativeclient\cppcache\src\TcpConn.cpp)
  c:\nativeclient\cppcache\src\TcpConn.hpp(33): note: see declaration of 'apache::geode::client::TcpConn' (compiling source file C:\nativeclient\cppcache\src\TcpConn.cpp)
c:\nativeclient\cppcache\src\TcpConn.hpp(33): error C2220: warning treated as error - no 'object' file generated (compiling source file C:\nativeclient\cppcache\src\TcpSslConn.cpp) [C:\build\cppcache\shared\apache-geode.vcxproj]
c:\nativeclient\cppcache\src\TcpConn.hpp(33): warning C4275: non dll-interface class 'apache::geode::client::Connector' used as base for dll-interface class 'apache::geode::client::TcpConn' (compiling source file C:\nativeclient\cppcache\src\TcpSslConn.cpp) [C:\build\cppcache\shared\apache-geode.vcxproj]
  c:\nativeclient\cppcache\src\Connector.hpp(40): note: see declaration of 'apache::geode::client::Connector' (compiling source file C:\nativeclient\
```

It looks like TcpConn has also been exported and wasn't before.  Perhaps fixing that will take care of everything.

**Comment by albertogpz** [comment]
> Well I guess we know why Connector needs to be APACHE_GEODE_EXPORT now:
> 
> ```
> c:\nativeclient\cppcache\src\TcpConn.hpp(33): error C2220: warning treated as error - no 'object' file generated (compiling source file C:\nativeclient\cppcache\src\TcpConn.cpp) [C:\build\cppcache\shared\apache-geode.vcxproj]
> c:\nativeclient\cppcache\src\TcpConn.hpp(33): warning C4275: non dll-interface class 'apache::geode::client::Connector' used as base for dll-interface class 'apache::geode::client::TcpConn' (compiling source file C:\nativeclient\cppcache\src\TcpConn.cpp) [C:\build\cppcache\shared\apache-geode.vcxproj]
>   c:\nativeclient\cppcache\src\Connector.hpp(40): note: see declaration of 'apache::geode::client::Connector' (compiling source file C:\nativeclient\cppcache\src\TcpConn.cpp)
>   c:\nativeclient\cppcache\src\TcpConn.hpp(33): note: see declaration of 'apache::geode::client::TcpConn' (compiling source file C:\nativeclient\cppcache\src\TcpConn.cpp)
> c:\nativeclient\cppcache\src\TcpConn.hpp(33): error C2220: warning treated as error - no 'object' file generated (compiling source file C:\nativeclient\cppcache\src\TcpSslConn.cpp) [C:\build\cppcache\shared\apache-geode.vcxproj]
> c:\nativeclient\cppcache\src\TcpConn.hpp(33): warning C4275: non dll-interface class 'apache::geode::client::Connector' used as base for dll-interface class 'apache::geode::client::TcpConn' (compiling source file C:\nativeclient\cppcache\src\TcpSslConn.cpp) [C:\build\cppcache\shared\apache-geode.vcxproj]
>   c:\nativeclient\cppcache\src\Connector.hpp(40): note: see declaration of 'apache::geode::client::Connector' (compiling source file C:\nativeclient\
> ```
> 
> It looks like TcpConn has also been exported and wasn't before. Perhaps fixing that will take care of everything.

Let's see.

**Comment by pdxcodemonkey** [comment]
Confirmed - removing APACHE_GEODE_EXPORT from TcpConn fixes my Windows build.

File: defaultSystem/geode.properties (hunk 5 above)
[the remark this answers is not in the document]
Comment (pdxcodemonkey) [reply defaultSystem/geode.properties]: Okay - we can clean this up later.

File: cppcache/src/TcrConnection.hpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (pdxcodemonkey) [reply cppcache/src/TcrConnection.hpp:112]: I suppose thanks are due to @mreddington then :).  Either way, this is much better.

**Review by pdxcodemonkey — APPROVED** [review]

**Pull request merged**